File: src/test/java/com/joecollins/graphics/components/HeatMapFrameBuilderTest.java
package com.joecollins.graphics.components;

import static java.awt.Color.BLUE;
import static java.awt.Color.GREEN;
import static java.awt.Color.RED;
import static org.junit.Assert.assertEquals;

import com.joecollins.bindings.Bindable;
import com.joecollins.bindings.BindableList;
import com.joecollins.bindings.Binding;
import com.joecollins.graphics.utils.ColorUtils;
import com.joecollins.models.general.Party;
import com.joecollins.models.general.PartyResult;
import java.awt.Color;
import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import org.apache.commons.lang3.tuple.ImmutablePair;
import org.apache.commons.lang3.tuple.Pair;
import org.junit.Test;

public class HeatMapFrameBuilderTest {

  @Test
  public void testHeatMapBasic() {
    List<Pair<Color, Color>> dots = new ArrayList<>();
    dots.addAll(Collections.nCopies(1, ImmutablePair.of(GREEN, GREEN)));
    dots.addAll(Collections.nCopies(7, ImmutablePair.of(GREEN, RED)));
    dots.addAll(Collections.nCopies(6, ImmutablePair.of(RED, RED)));
    dots.addAll(Collections.nCopies(5, ImmutablePair.of(BLUE, RED)));
    dots.addAll(Collections.nCopies(8, ImmutablePair.of(BLUE, BLUE)));

    BindableList<Pair<Color, Integer>> seatBars =
        new BindableList<>(List.of(ImmutablePair.of(GREEN, 8)));
    BindableList<Pair<Color, Integer>> changeBars =
        new BindableList<>(List.of(ImmutablePair.of(GREEN, +7)));

    HeatMapFrame frame =
        HeatMapFrameBuilder.of(
                () -> 3,
                dots,
                e -> Binding.fixedBinding(e.getLeft()),
                e -> Binding.fixedBinding(e.getRight()))
            .withSeatBars(seatBars, Pair::getLeft, Pair::getRight, Binding.fixedBinding("GREEN: 8"))
            .withChangeBars(
                changeBars,
                Pair::getLeft,
                Pair::getRight,
                Binding.fixedBinding(1),
                Binding.fixedBinding("GRN: +7"))
            .withHeader(() -> "PEI")
            .withBorder(() -> GREEN)
            .build();

    assertEquals(3, frame.getNumRows());

    assertEquals(27, frame.getNumSquares());
    List<Color> expectedFills =
        Stream.of(
                Collections.nCopies(8, GREEN),
                Collections.nCopies(6, RED),
                Collections.nCopies(13, BLUE))
            .flatMap(Collection::stream)
            .collect(Collectors.toList());
    List<Color> expectedBorders =
        Stream.of(
                Collections.nCopies(1, GREEN),
                Collections.nCopies(18, RED),
                Collections.nCopies(8, BLUE))
            .flatMap(Collection::stream)
            .collect(Collectors.toList());
    for (int i = 0; i < frame.getNumSquares(); i++) {
      assertEquals("Square fill " + i, expectedFills.get(i), frame.getSquareFill(i));
      assertEquals("Square border " + i, expectedBorders.get(i), frame.getSquareBorder(i));
    }

    assertEquals(1, frame.getSeatBarCount());
    assertEquals(GREEN, frame.getSeatBarColor(0));
    assertEquals(8, frame.getSeatBarSize(0));
    assertEquals("GREEN: 8", frame.getSeatBarLabel());

    assertEquals(1, frame.getChangeBarCount());
    assertEquals(GREEN, frame.getChangeBarColor(0));
    assertEquals(7, frame.getChangeBarSize(0));
    assertEquals(1, frame.getChangeBarStart());
    assertEquals("GRN: +7", frame.getChangeBarLabel());

    assertEquals("PEI", frame.getHeader());
    assertEquals(GREEN, frame.getBorderColor());
  }

  @Test
  public void testBasicLeadingElected() {
    Party lib = new Party("Liberal", "LIB", Color.RED);
    Party yp = new Party("Yukon Party", "YP", Color.BLUE);
    Party ndp = new Party("New Democratic Party", "NDP", Color.ORANGE);

    class Riding {
      final String name;
      final Party leader;
      final boolean hasWon;
      final Party prev;

      Riding(String name, Party leader, boolean hasWon, Party prev) {
        this.name = name;
        this.leader = leader;
        this.hasWon = hasWon;
        this.prev = prev;
      }
    }

    List<Riding> ridings = new ArrayList<>();
    ridings.add(new Riding("Vuntut Gwitchin", lib, false, lib));
    ridings.add(new Riding("Klondike", lib, true, lib));
    ridings.add(new Riding("Takhini-Copper King", ndp, false, ndp));
    ridings.add(new Riding("Whitehorse Centre", ndp, false, ndp));
    ridings.add(new Riding("Mayo-Tatchun", lib, true, ndp));
    ridings.add(new Riding("Mount Lorne-Southern Lakes", lib, false, ndp));
    ridings.add(new Riding("Riverdale South", lib, false, ndp));
    ridings.add(new Riding("Copperbelt South", yp, false, ndp));
    ridings.add(new Riding("Porter Creek South", lib, false, yp));
    ridings.add(new Riding("Watson Lake", yp, true, yp));
    ridings.add(new Riding("Porter Creek Centre", lib, false, yp));
    ridings.add(new Riding("Riverdale North", lib, true, yp));
    ridings.add(new Riding("Kluane", yp, false, yp));
    ridings.add(new Riding("Mountainview", lib, false, yp));
    ridings.add(new Riding("Copperbelt North", lib, false, yp));
    ridings.add(new Riding("Pelly-Nisutlin", yp, true, yp));
    ridings.add(new Riding("Porter Creek North", yp, false, yp));
    ridings.add(new Riding("Lake Laberge", yp, true, yp));
    ridings.add(new Riding("Whitehorse West", lib, false, yp));

    HeatMapFrame frame =
        HeatMapFrameBuilder.ofElectedLeading(
            () -> 3,
            ridings,
            r -> Binding.fixedBinding(new PartyResult(r.leader, r.hasWon)),
            r -> r.prev,
            lib,
            (e, l) -> "LIB: " + e + "/" + l,
            (e, l) -> l > 0,
            (e, l) ->
                new DecimalFormat("+0;-0").format(e) + "/" + new DecimalFormat("+0;-0").format(l),
            () -> "YUKON");

    assertEquals(19, frame.getNumSquares());
    assertEquals(Color.RED, frame.getSquareBorder(0));
    assertEquals(ColorUtils.lighten(Color.RED), frame.getSquareFill(0));
    assertEquals("YUKON", frame.getHeader());
    assertEquals(Color.RED, frame.getBorderColor());

    assertEquals(2, frame.getSeatBarCount());
    assertEquals(Color.RED, frame.getSeatBarColor(0));
    assertEquals(3, frame.getSeatBarSize(0));
    assertEquals(ColorUtils.lighten(Color.RED), frame.getSeatBarColor(1));
    assertEquals(8, frame.getSeatBarSize(1));
    assertEquals("LIB: 3/11", frame.getSeatBarLabel());

    assertEquals(2, frame.getChangeBarStart());
    assertEquals(2, frame.getChangeBarCount());
    assertEquals(Color.RED, frame.getChangeBarColor(0));
    assertEquals(2, frame.getChangeBarSize(0));
    assertEquals(ColorUtils.lighten(Color.RED), frame.getChangeBarColor(1));
    assertEquals(7, frame.getChangeBarSize(1));
    assertEquals("+2/+9", frame.getChangeBarLabel());
  }

  enum Property {
    PROP
  }

  @Test
  public void testLeadingElectedSeatsRepeated() {
    Party dem = new Party("Democratic", "DEM", Color.BLUE);
    Party gop = new Party("Republican", "GOP", Color.RED);

    class Result extends Bindable<Property> {
      Party leader;
      boolean hasWon;
      final Party prev;

      Result(Party leader, boolean hasWon, Party prev) {
        this.leader = leader;
        this.hasWon = hasWon;
        this.prev = prev;
      }

      Binding<PartyResult> getBinding() {
        return Binding.propertyBinding(
            this,
            t -> {
              if (leader == null) return null;
              return new PartyResult(leader, hasWon);
            },
            Property.PROP);
      }

      void setResult(Party leader, boolean hasWon) {
        this.leader = leader;
        this.hasWon = hasWon;
        onPropertyRefreshed(Property.PROP);
      }
    }

    Result result = new Result(null, false, gop);
    List<Result> results = Collections.nCopies(30, result);
    HeatMapFrame frame =
        HeatMapFrameBuilder.ofElectedLeading(
            Binding.fixedBinding(results.size()),
            results,
            r -> r.getBinding(),
            r -> r.prev,
            dem,
            (e, l) -> "DEM: " + e + "/" + l,
            (e, l) -> l > 0,
            (e, l) ->
                new DecimalFormat("+0;-0").format(e) + "/" + new DecimalFormat("+0;-0").format(l),
            () -> "TEST");
    assertEquals(30, frame.getNumSquares());
    assertEquals(Color.RED, frame.getSquareBorder(0));
    assertEquals(Color.WHITE, frame.getSquareFill(0));
    assertEquals(0, frame.getSeatBarSize(0));
    assertEquals(0, frame.getSeatBarSize(1));

    result.setResult(gop, false);
    assertEquals(Color.RED, frame.getSquareBorder(0));
    assertEquals(ColorUtils.lighten(Color.RED), frame.getSquareFill(0));
    assertEquals(0, frame.getSeatBarSize(0));
    assertEquals(0, frame.getSeatBarSize(1));

    result.setResult(dem, true);
    assertEquals(Color.RED, frame.getSquareBorder(0));
    assertEquals(Color.BLUE, frame.getSquareFill(0));
    assertEquals(30, frame.getSeatBarSize(0));
    assertEquals(0, frame.getSeatBarSize(1));
  }

  @Test
  public void testLeadingElectedMultiseat() {
    Party dem = new Party("Democratic", "DEM", Color.BLUE);
    Party gop = new Party("Republican", "GOP", Color.RED);

    class Result extends Bindable<Property> {
      Party leader;
      boolean hasWon;
      final Party prev;
      final int numSeats;

      Result(Party leader, boolean hasWon, Party prev, int numSeats) {
        this.leader = leader;
        this.hasWon = hasWon;
        this.prev = prev;
        this.numSeats = numSeats;
      }

      Binding<PartyResult> getBinding() {
        return Binding.propertyBinding(
            this,
            t -> {
              if (leader == null) return null;
              return new PartyResult(leader, hasWon);
            },
            Property.PROP);
      }

      void setResult(Party leader, boolean hasWon) {
        this.leader = leader;
        this.hasWon = hasWon;
        onPropertyRefreshed(Property.PROP);
      }
    }

    Result result = new Result(null, false, gop, 30);
    List<Result> results = Arrays.asList(result);
    HeatMapFrame frame =
        HeatMapFrameBuilder.ofElectedLeading(
            Binding.fixedBinding(results.stream().mapToInt(r -> r.numSeats).sum()),
            results,
            r -> r.numSeats,
            r -> r.getBinding(),
            r -> r.prev,
            dem,
            (e, l) -> "DEM: " + e + "/" + l,
            (e, l) -> true,
            (e, l) ->
                new DecimalFormat("+0;-0").format(e) + "/" + new DecimalFormat("+0;-0").format(l),
            () -> "TEST");
    assertEquals(30, frame.getNumSquares());
    assertEquals(Color.RED, frame.getSquareBorder(0));
    assertEquals(Color.WHITE, frame.getSquareFill(0));
    assertEquals(0, frame.getSeatBarSize(0));
    assertEquals(0, frame.getSeatBarSize(1));
    assertEquals(0, frame.getChangeBarSize(0));
    assertEquals(0, frame.getChangeBarSize(1));

    result.setResult(gop, false);
    assertEquals(Color.RED, frame.getSquareBorder(0));
    assertEquals(ColorUtils.lighten(Color.RED), frame.getSquareFill(0));
    assertEquals(0, frame.getSeatBarSize(0));
    assertEquals(0, frame.getSeatBarSize(1));
    assertEquals(0, frame.getChangeBarSize(0));
    assertEquals(0, frame.getChangeBarSize(1));

    result.setResult(dem, false);
    assertEquals(Color.RED, frame.getSquareBorder(0));
    assertEquals(ColorUtils.lighten(Color.BLUE), frame.getSquareFill(0));
    assertEquals(0, frame.getSeatBarSize(0));
    assertEquals(30, frame.getSeatBarSize(1));
    assertEquals(0, frame.getChangeBarSize(0));
    assertEquals(30, frame.getChangeBarSize(1));

    result.setResult(dem, true);
    assertEquals(Color.RED, frame.getSquareBorder(0));
    assertEquals(Color.BLUE, frame.getSquareFill(0));
    assertEquals(30, frame.getSeatBarSize(0));
    assertEquals(0, frame.getSeatBarSize(1));
    assertEquals(30, frame.getChangeBarSize(0));
    assertEquals(0, frame.getChangeBarSize(1));
  }

  @Test
  public void testBarsUpdateProperlyOnIncompleteResults() {
    Party dem = new Party("Democratic", "DEM", Color.BLUE);
    Party gop = new Party("Republican", "GOP", Color.RED);

    class Result extends Bindable<Property> {
      Party leader;
      boolean hasWon;
      final Party prev;
      final int numSeats;

      Result(Party leader, boolean hasWon, Party prev, int numSeats) {
        this.leader = leader;
        this.hasWon = hasWon;
        this.prev = prev;
        this.numSeats = numSeats;
      }

      Binding<PartyResult> getBinding() {
        return Binding.propertyBinding(this, t -> new PartyResult(leader, hasWon), Property.PROP);
      }

      void setResult(Party leader, boolean hasWon) {
        this.leader = leader;
        this.hasWon = hasWon;
        onPropertyRefreshed(Property.PROP);
      }
    }

    Result result = new Result(null, false, gop, 30);
    List<Result> results = Arrays.asList(result);
    HeatMapFrame frame =
        HeatMapFrameBuilder.ofElectedLeading(
            Binding.fixedBinding(results.stream().mapToInt(r -> r.numSeats).sum()),
            results,
            r -> r.numSeats,
            r -> r.getBinding(),
            r -> r.prev,
            dem,
            (e, l) -> "DEM: " + e + "/" + l,
            (e, l) -> true,
            (e, l) ->
                new DecimalFormat("+0;-0").format(e) + "/" + new DecimalFormat("+0;-0").format(l),
            () -> "TEST");
    assertEquals(30, frame.getNumSquares());
    assertEquals(Color.RED, frame.getSquareBorder(0));
    assertEquals(Color.WHITE, frame.getSquareFill(0));
    assertEquals(0, frame.getSeatBarSize(0));
    assertEquals(0, frame.getSeatBarSize(1));
    assertEquals(0, frame.getChangeBarSize(0));
    assertEquals(0, frame.getChangeBarSize(1));

    result.setResult(null, false);
    assertEquals(Color.RED, frame.getSquareBorder(0));
    assertEquals(Color.WHITE, frame.getSquareFill(0));
    assertEquals(0, frame.getSeatBarSize(0));
    assertEquals(0, frame.getSeatBarSize(1));
    assertEquals(0, frame.getChangeBarSize(0));
    assertEquals(0, frame.getChangeBarSize(1));
  }
}

File: src/test/java/com/joecollins/graphics/components/HeatMapFrameBuilderTest.kt
package com.joecollins.graphics.components

import com.joecollins.bindings.Bindable
import com.joecollins.bindings.Binding
import com.joecollins.bindings.Binding.Companion.fixedBinding
import com.joecollins.bindings.Binding.Companion.propertyBinding
import com.joecollins.graphics.components.HeatMapFrameBuilder.Companion.of
import com.joecollins.graphics.components.HeatMapFrameBuilder.Companion.ofElectedLeading
import com.joecollins.graphics.utils.BindableWrapper
import com.joecollins.graphics.utils.ColorUtils.lighten
import com.joecollins.models.general.Party
import com.joecollins.models.general.PartyResult
import org.awaitility.Awaitility
import org.hamcrest.core.IsEqual
import org.junit.Assert
import org.junit.Test
import java.awt.Color
import java.text.DecimalFormat
import java.util.ArrayList
import java.util.Collections
import java.util.concurrent.TimeUnit

class HeatMapFrameBuilderTest {
    @Test
    fun testHeatMapBasic() {
        val dots: MutableList<Pair<Color, Color>> = ArrayList()
        dots.addAll(Collections.nCopies(1, Pair(Color.GREEN, Color.GREEN)))
        dots.addAll(Collections.nCopies(7, Pair(Color.GREEN, Color.RED)))
        dots.addAll(Collections.nCopies(6, Pair(Color.RED, Color.RED)))
        dots.addAll(Collections.nCopies(5, Pair(Color.BLUE, Color.RED)))
        dots.addAll(Collections.nCopies(8, Pair(Color.BLUE, Color.BLUE)))
        val seatBars = BindableWrapper(listOf(Pair(Color.GREEN, 8)))
        val changeBars = BindableWrapper(listOf(Pair(Color.GREEN, +7)))
        val frame = of(
            fixedBinding(3),
            dots,
            { fixedBinding(it.first) },
            { fixedBinding(it.second) },
            { fixedBinding(null) }
        )
            .withSeatBars(seatBars.binding, { it.first }, { it.second }, fixedBinding("GREEN: 8"))
            .withChangeBars(
                changeBars.binding, { it.first }, { it.second },
                fixedBinding(1),
                fixedBinding("GRN: +7")
            )
            .withHeader(fixedBinding("PEI"))
            .withBorder(fixedBinding(Color.GREEN))
            .build()
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ frame.numRows }, IsEqual(3))
        Assert.assertEquals(27, frame.numSquares.toLong())
        val expectedFills = sequenceOf(
            Collections.nCopies(8, Color.GREEN),
            Collections.nCopies(6, Color.RED),
            Collections.nCopies(13, Color.BLUE)
        )
            .flatten()
            .toList()
        val expectedBorders = sequenceOf(
            Collections.nCopies(1, Color.GREEN),
            Collections.nCopies(18, Color.RED),
            Collections.nCopies(8, Color.BLUE)
        )
            .flatten()
            .toList()
        for (i in 0 until frame.numSquares) {
            Assert.assertEquals("Square fill $i", expectedFills[i], frame.getSquareFill(i))
            Assert.assertEquals("Square border $i", expectedBorders[i], frame.getSquareBorder(i))
        }
        Assert.assertEquals(1, frame.seatBarCount.toLong())
        Assert.assertEquals(Color.GREEN, frame.getSeatBarColor(0))
        Assert.assertEquals(8, frame.getSeatBarSize(0).toLong())
        Assert.assertEquals("GREEN: 8", frame.seatBarLabel)
        Assert.assertEquals(1, frame.changeBarCount.toLong())
        Assert.assertEquals(Color.GREEN, frame.getChangeBarColor(0))
        Assert.assertEquals(7, frame.getChangeBarSize(0).toLong())
        Assert.assertEquals(1, frame.changeBarStart.toLong())
        Assert.assertEquals("GRN: +7", frame.changeBarLabel)
        Assert.assertEquals("PEI", frame.header)
        Assert.assertEquals(Color.GREEN, frame.borderColor)
    }

    @Test
    fun testBasicLeadingElected() {
        val lib = Party("Liberal", "LIB", Color.RED)
        val yp = Party("Yukon Party", "YP", Color.BLUE)
        val ndp = Party("New Democratic Party", "NDP", Color.ORANGE)

        class Riding(val name: String, val leader: Party, val hasWon: Boolean, val prev: Party)

        val ridings: MutableList<Riding> = ArrayList()
        ridings.add(Riding("Vuntut Gwitchin", lib, false, lib))
        ridings.add(Riding("Klondike", lib, true, lib))
        ridings.add(Riding("Takhini-Copper King", ndp, false, ndp))
        ridings.add(Riding("Whitehorse Centre", ndp, false, ndp))
        ridings.add(Riding("Mayo-Tatchun", lib, true, ndp))
        ridings.add(Riding("Mount Lorne-Southern Lakes", lib, false, ndp))
        ridings.add(Riding("Riverdale South", lib, false, ndp))
        ridings.add(Riding("Copperbelt South", yp, false, ndp))
        ridings.add(Riding("Porter Creek South", lib, false, yp))
        ridings.add(Riding("Watson Lake", yp, true, yp))
        ridings.add(Riding("Porter Creek Centre", lib, false, yp))
        ridings.add(Riding("Riverdale North", lib, true, yp))
        ridings.add(Riding("Kluane", yp, false, yp))
        ridings.add(Riding("Mountainview", lib, false, yp))
        ridings.add(Riding("Copperbelt North", lib, false, yp))
        ridings.add(Riding("Pelly-Nisutlin", yp, true, yp))
        ridings.add(Riding("Porter Creek North", yp, false, yp))
        ridings.add(Riding("Lake Laberge", yp, true, yp))
        ridings.add(Riding("Whitehorse West", lib, false, yp))
        val frame = ofElectedLeading(
            fixedBinding(3),
            ridings,
            { fixedBinding(PartyResult(it.leader, it.hasWon)) },
            { it.prev },
            lib,
            { e: Int, l: Int -> "LIB: $e/$l" },
            { _: Int, l: Int -> l > 0 },
            { e: Int, l: Int -> DecimalFormat("+0;-0").format(e) + "/" + DecimalFormat("+0;-0").format(l) },
            fixedBinding("YUKON")
        )
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ frame.numSquares }, IsEqual(19))
        Assert.assertEquals(Color.RED, frame.getSquareBorder(0))
        Assert.assertEquals(lighten(Color.RED), frame.getSquareFill(0))
        Assert.assertEquals("YUKON", frame.header)
        Assert.assertEquals(Color.RED, frame.borderColor)
        Assert.assertEquals(2, frame.seatBarCount.toLong())
        Assert.assertEquals(Color.RED, frame.getSeatBarColor(0))
        Assert.assertEquals(3, frame.getSeatBarSize(0).toLong())
        Assert.assertEquals(lighten(Color.RED), frame.getSeatBarColor(1))
        Assert.assertEquals(8, frame.getSeatBarSize(1).toLong())
        Assert.assertEquals("LIB: 3/11", frame.seatBarLabel)
        Assert.assertEquals(2, frame.changeBarStart.toLong())
        Assert.assertEquals(2, frame.changeBarCount.toLong())
        Assert.assertEquals(Color.RED, frame.getChangeBarColor(0))
        Assert.assertEquals(2, frame.getChangeBarSize(0).toLong())
        Assert.assertEquals(lighten(Color.RED), frame.getChangeBarColor(1))
        Assert.assertEquals(7, frame.getChangeBarSize(1).toLong())
        Assert.assertEquals("+2/+9", frame.changeBarLabel)
    }

    internal enum class Property {
        PROP
    }

    @Test
    fun testLeadingElectedSeatsRepeated() {
        val dem = Party("Democratic", "DEM", Color.BLUE)
        val gop = Party("Republican", "GOP", Color.RED)

        class Result(var leader: Party?, var hasWon: Boolean, val prev: Party) : Bindable<Result, Property>() {
            val binding: Binding<PartyResult?>
                get() = propertyBinding(
                    this,
                    {
                        if (leader == null) null
                        else PartyResult(leader, hasWon)
                    },
                    Property.PROP
                )

            fun setResult(leader: Party?, hasWon: Boolean) {
                this.leader = leader
                this.hasWon = hasWon
                onPropertyRefreshed(Property.PROP)
            }
        }

        val result = Result(null, false, gop)
        val results = Collections.nCopies(30, result)
        val frame = ofElectedLeading(
            fixedBinding(results.size),
            results,
            { it.binding },
            { it.prev },
            dem,
            { e: Int, l: Int -> "DEM: $e/$l" },
            { _: Int, l: Int -> l > 0 },
            { e: Int, l: Int -> DecimalFormat("+0;-0").format(e) + "/" + DecimalFormat("+0;-0").format(l) },
            fixedBinding("TEST")
        )
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ frame.numSquares }, IsEqual(30))
        Assert.assertEquals(Color.RED, frame.getSquareBorder(0))
        Assert.assertEquals(Color.WHITE, frame.getSquareFill(0))
        Assert.assertEquals(0, frame.getSeatBarSize(0).toLong())
        Assert.assertEquals(0, frame.getSeatBarSize(1).toLong())
        result.setResult(gop, false)
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ frame.getSquareFill(0) }, IsEqual(lighten(Color.RED)))
        Assert.assertEquals(Color.RED, frame.getSquareBorder(0))
        Assert.assertEquals(lighten(Color.RED), frame.getSquareFill(0))
        Assert.assertEquals(0, frame.getSeatBarSize(0).toLong())
        Assert.assertEquals(0, frame.getSeatBarSize(1).toLong())
        result.setResult(dem, true)
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ frame.getSquareFill(0) }, IsEqual(Color.BLUE))
        Assert.assertEquals(Color.RED, frame.getSquareBorder(0))
        Assert.assertEquals(Color.BLUE, frame.getSquareFill(0))
        Assert.assertEquals(30, frame.getSeatBarSize(0).toLong())
        Assert.assertEquals(0, frame.getSeatBarSize(1).toLong())
    }

    @Test
    fun testLeadingElectedMultiseat() {
        val dem = Party("Democratic", "DEM", Color.BLUE)
        val gop = Party("Republican", "GOP", Color.RED)

        class Result(var leader: Party?, var hasWon: Boolean, val prev: Party, val numSeats: Int) : Bindable<Result, Property>() {
            val binding: Binding<PartyResult?>
                get() = propertyBinding(
                    this,
                    {
                        if (leader == null) null
                        else PartyResult(leader, hasWon)
                    },
                    Property.PROP
                )

            fun setResult(leader: Party?, hasWon: Boolean) {
                this.leader = leader
                this.hasWon = hasWon
                onPropertyRefreshed(Property.PROP)
            }
        }

        val result = Result(null, false, gop, 30)
        val results = listOf(result)
        val frame = ofElectedLeading(
            fixedBinding(results.map { it.numSeats }.sum()),
            results,
            { it.numSeats },
            { it.binding },
            { it.prev },
            dem,
            { e: Int, l: Int -> "DEM: $e/$l" },
            { _: Int, _: Int -> true },
            { e: Int, l: Int -> DecimalFormat("+0;-0").format(e) + "/" + DecimalFormat("+0;-0").format(l) },
            fixedBinding("TEST")
        )
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ frame.numSquares }, IsEqual(30))
        Assert.assertEquals(Color.RED, frame.getSquareBorder(0))
        Assert.assertEquals(Color.WHITE, frame.getSquareFill(0))
        Assert.assertEquals(0, frame.getSeatBarSize(0).toLong())
        Assert.assertEquals(0, frame.getSeatBarSize(1).toLong())
        Assert.assertEquals(0, frame.getChangeBarSize(0).toLong())
        Assert.assertEquals(0, frame.getChangeBarSize(1).toLong())
        result.setResult(gop, false)
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ frame.getSquareFill(0) }, IsEqual(lighten(Color.RED)))
        Assert.assertEquals(Color.RED, frame.getSquareBorder(0))
        Assert.assertEquals(lighten(Color.RED), frame.getSquareFill(0))
        Assert.assertEquals(0, frame.getSeatBarSize(0).toLong())
        Assert.assertEquals(0, frame.getSeatBarSize(1).toLong())
        Assert.assertEquals(0, frame.getChangeBarSize(0).toLong())
        Assert.assertEquals(0, frame.getChangeBarSize(1).toLong())
        result.setResult(dem, false)
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ frame.getSquareFill(0) }, IsEqual(lighten(Color.BLUE)))
        Assert.assertEquals(Color.RED, frame.getSquareBorder(0))
        Assert.assertEquals(0, frame.getSeatBarSize(0).toLong())
        Assert.assertEquals(30, frame.getSeatBarSize(1).toLong())
        Assert.assertEquals(0, frame.getChangeBarSize(0).toLong())
        Assert.assertEquals(30, frame.getChangeBarSize(1).toLong())
        result.setResult(dem, true)
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ frame.getSquareFill(0) }, IsEqual(Color.BLUE))
        Assert.assertEquals(Color.RED, frame.getSquareBorder(0))
        Assert.assertEquals(30, frame.getSeatBarSize(0).toLong())
        Assert.assertEquals(0, frame.getSeatBarSize(1).toLong())
        Assert.assertEquals(30, frame.getChangeBarSize(0).toLong())
        Assert.assertEquals(0, frame.getChangeBarSize(1).toLong())
    }

    @Test
    fun testBarsUpdateProperlyOnIncompleteResults() {
        val dem = Party("Democratic", "DEM", Color.BLUE)
        val gop = Party("Republican", "GOP", Color.RED)

        class Result(var leader: Party?, var hasWon: Boolean, val prev: Party, val numSeats: Int) : Bindable<Result, Property>() {
            val binding: Binding<PartyResult?>
                get() = propertyBinding(this, { PartyResult(leader, hasWon) }, Property.PROP)

            fun setResult(leader: Party?, hasWon: Boolean) {
                this.leader = leader
                this.hasWon = hasWon
                onPropertyRefreshed(Property.PROP)
            }
        }

        val result = Result(null, false, gop, 30)
        val results = listOf(result)
        val frame = ofElectedLeading(
            fixedBinding(results.map { it.numSeats }.sum()),
            results,
            { it.numSeats },
            { it.binding },
            { it.prev },
            dem,
            { e: Int, l: Int -> "DEM: $e/$l" },
            { _: Int, _: Int -> true },
            { e: Int, l: Int -> DecimalFormat("+0;-0").format(e) + "/" + DecimalFormat("+0;-0").format(l) },
            fixedBinding("TEST")
        )
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ frame.numSquares }, IsEqual(30))
        Assert.assertEquals(Color.RED, frame.getSquareBorder(0))
        Assert.assertEquals(Color.WHITE, frame.getSquareFill(0))
        Assert.assertEquals(0, frame.getSeatBarSize(0).toLong())
        Assert.assertEquals(0, frame.getSeatBarSize(1).toLong())
        Assert.assertEquals(0, frame.getChangeBarSize(0).toLong())
        Assert.assertEquals(0, frame.getChangeBarSize(1).toLong())
        result.setResult(null, false)
        Assert.assertEquals(Color.RED, frame.getSquareBorder(0))
        Assert.assertEquals(Color.WHITE, frame.getSquareFill(0))
        Assert.assertEquals(0, frame.getSeatBarSize(0).toLong())
        Assert.assertEquals(0, frame.getSeatBarSize(1).toLong())
        Assert.assertEquals(0, frame.getChangeBarSize(0).toLong())
        Assert.assertEquals(0, frame.getChangeBarSize(1).toLong())
    }
}
